### DISCUSSION TITLE
Eigene IDs im Interface vergeben

### DISCUSSION THREAD

**Discussion by fb4thxUall** [jzakotnik/openlibry#95] [Ideas]
Lieber Jure,
hallo zusammen,

ich schlage auch hier nochmal öffentlich diskutierbar folgendes neues Feature vor: Man sollte beim Neuanlegen eines Bib-Users auch eine eigene ID im grafischen Interface vergeben können. Das hat zum Beispiel für Schulen den Vorteil, dass man bestehende IDs aus bereits existierenden Systemen (Schulverwaltungssoftware, Mensasystem usw.) auch in openlibry verwenden kann. Das wiederum ermöglicht es, dass man nicht alle persönlichen Daten für alle Bib-Betreuer (bei uns sind das auch ältere SchülerInnen) sichtbar in OpenLibry hinterlegen muss. Es genügt zum Beispiel der Name und die (fremde) ID. Sollte dann Bedarf bestehen, die Daten zu erfahren, kann die verantwortliche Lehrkraft im Schulsystem unter der ID nachschlagen. 
Das erspart einige Datenschutzprobleme.

Bei uns ist diese ID auch auf dem Schülerausweis abgedruckt, sodass dieser automatisch mit als Bib-Ausweis fungieren kann. 

Gleichzeitig hätte das Feature keine Einschänkungen für andere Use-Cases, da ja auch immer noch eine auto-generierte ID optional genutzt werden kann. Auch ist der Implementationsaufwand mMn überschaubar. 

Für meine Schule wäre das der letzte verbleibende entscheidende Game-Changer, um openlibry einzuführen. 

Ganz herzlichen Dank insbesondere an Jure für die tolle Arbeit.

Viele Grüße

**Comment by jzakotnik** [comment]
@fb4thxUall Danke für das Feedback! Glaube der einfachste Weg wäre bei der Anlage des Users die ID abzufragen - danach ist sie aber fix, sonst wird es zu kompliziert z.B. die Leihen upzudaten. Sollte in den nächsten Tagen im Main Branch sein 💯 .

**Reply by fb4thxUall** [reply]
Ja natürlich, die ID bleibt anschließend ein Schülerleben lang. Anders wäre schwierig. Sehr sehr cool. Dankeschön.

**Comment by jzakotnik** [comment]
Almost there, nur noch ein wenig Fehlerhandling falls die ID schon existiert.. Ansonsten spuckt die automatische ID die nächste aus, welche die Datenbank vergeben würde, also @fb4thxUall falls ihr die IDs hintereinander eingebt, dann müsstet ihr theoretisch gar nicht viel manuell eingeben?
<img width="693" alt="image" src="https://github.com/jzakotnik/openlibry/assets/551599/d2deb165-0127-4e99-bf90-213297638c01">

**Reply by fb4thxUall** [reply]
Das klingt praktisch, ist bei uns konkret aber tatsächlich nicht der Fall. Da sind große Lücken. Aber eine gute Idee, hilft anderen vielleicht.

**Comment by jzakotnik** [comment]
@fb4thxUall die erste Version für manuelle IDs ist jetzt im main branch, magst es mal ausprobieren?

**Reply by fb4thxUall** [reply]
Hallo Jure,

ich komme endlich dazu, es zu testen. Vielen Dank für das Feature. Ich werde berichten. Je länger meine Antwort dauert, desto besser, denn dann läuft es gut. Ich melde mich aber so oder so. Liebe Grüße

**Reply by jzakotnik** [reply]
Viel Spaß, falls es nicht klappt sag Bescheid 💯

**Reply by fb4thxUall** [reply]
Hi,

es gibt tatsächlich direkt 2 Kleinigkeiten:

1.: Für die ID wäre BIGINT ausfallsicherer. Direkt meine erste (erfundene) Test-ID sprengte INT und ich weiß gerade nicht sicher, ob die realen IDs kurz genug sind. Das Problem kommt natürlich nur das das selbst Vergeben, "normales" INT ist natürlich die nachvollziehbare Wahl. Man kann im Feld eine größere Zahl eingeben. Allerdings stürzt dann das Programm ab. (Fehler 500: Internal Server Error). Docker gibt in der Konsole auch an, dass es intern die Verwandlung in BIGINT versucht, dass es aber scheitert.

2.: Es wäre hilfreich, wenn "Klasse" kein Pflichtfeld wäre. Die Klasse ändert sich jedes Jahr. Ich würde gerne nur Vorname, Nachname und ID als Daten speichern, sonst nichts. Ich müsste ja sonst jedes Jahr die Klasse für alle Kinder aktualisieren. Es sollte mMn ein optionales Feld sein.


Weiter hab ich noch nicht getestet, muss gleich erstmal den Container neu aufsetzen, weil er durch den Int-Fehler nicht mehr zu gebrauchen ist und neu aufsetzen schneller geht als SQLite manuell befummeln. 

Apropos. Bevor ich ewig recherchiere. Hast du einen Link oder ein Stichwort für mich, wo ich möglichst schnell folgende Infos bekomme (ich arbeite leider das erste Mal mit Docker. Aber endlich, ich wollte mich schon immer mal einarbeiten...):

1. Wo finde ich die SQLite-Datei, damit ich ein Backup erstellen kann?
2. Wie mache ich auf einfachste Weise ein funktionierend wiederherstellbares Backup vom ganzen Container? Oder was ist eine kluge Strategie für schnelle Wiederherstellbarkeit im Falle eines GAUs? 

Falls du keine guten Links zur Hand hast: Kein Problem, ich werde mich selbst einfuchsen.


Ganz liebe Grüße und herzlichen Dank

**Reply by fb4thxUall** [reply]
Noch eine kleine Randbemerkung: Ich glaube, ein kurzer Leitfaden explizit für Nicht-Techies an Schulen mit Leitfaden von A-Z von Installation bis Instandhaltung wäre super praktisch. Ich werde damit mal anfangen und das zur Verfügung stellen, sobald hier alles dauerhaft funktionierend läuft. Ich hoffe, das ist in deinem Sinne?

**Reply by jzakotnik** [reply]
Danke!
1. Ich denke der Fehler ist ein anderer, weil in Node die maximale Zahl irgendwas bei 9 Quadrillionen ist. Welche Zahl hast Du ausprobiert? Würde es gern nachstellen..
2. Zum Glück ist die Klasse jetzt schon optional in der DB, das Frontend werde ich anpassen, gute Idee! #101 

Backup-Frage hatten auch andere :-) Es gibt zwei Optionen:
- Der Datenbankpfad ist im .env File unter `DATABASE_URL`, also z.B. `DATABASE_URL  = file:./somefolder/data.db `. Mit docker compose liegt die Datei auf einem persistent volume, siehe feld `volumes` im `docker-compose.yml`
- Alternativ kann man in der aktuellen Version ein excel downloaden und wieder hochladen. Das ist leichter verdaubar, verliert aber paar Daten (z.B. den Audit Log)

**Reply by jzakotnik** [reply]
> Noch eine kleine Randbemerkung: Ich glaube, ein kurzer Leitfaden explizit für Nicht-Techies an Schulen mit Leitfaden von A-Z von Installation bis Instandhaltung wäre super praktisch. Ich werde damit mal anfangen und das zur Verfügung stellen, sobald hier alles dauerhaft funktionierend läuft. Ich hoffe, das ist in deinem Sinne?

Na klar, je mehr Leute sich beteiligen, desto besser - das wäre ja der Vorteil von Open Source gegenüber kommerziellen Paketen..

**Reply by fb4thxUall** [reply]
Zum Fehler: Hier der komplette Log. Als Zusatzinfo: Ich habe ohne eingetragene Klasse auf Speichern geklickt.


`big@big-HP-250-G6-Notebook-PC:~/openlibry$ docker compose up
WARN[0000] /home/big/openlibry/docker-compose.yml: `version` is obsolete
[+] Running 1/0
 ✔ Container openlibry  Created                                                                                                                                                                                                          0.0s
Attaching to openlibry
openlibry  |
openlibry  | > [openlibry@0.9.0](mailto:openlibry@0.9.0) start
openlibry  | > PORT=3000 next start
openlibry  |
openlibry  |   ▲ Next.js 14.2.4
openlibry  |   - Local:        http://localhost:3000/
openlibry  |
openlibry  |  ✓ Starting...
openlibry  | Server startup
openlibry  |  ✓ Ready in 756ms
openlibry  | Server startup
openlibry  | ERROR in getAllUsers:  PrismaClientKnownRequestError:
openlibry  | Invalid `prisma.user.findMany()` invocation:
openlibry  |
openlibry  |
openlibry  | Inconsistent column data: Conversion failed: Value 18974892374 does not fit in an INT column, try migrating the 'id' column type to BIGINT
openlibry  |     at In.handleRequestError (/app/node_modules/@prisma/client/runtime/library.js:122:6877)
openlibry  |     at In.handleAndLogRequestError (/app/node_modules/@prisma/client/runtime/library.js:122:6211)
openlibry  |     at In.request (/app/node_modules/@prisma/client/runtime/library.js:122:5919)
openlibry  |     at async l (/app/node_modules/@prisma/client/runtime/library.js:127:11167)
openlibry  |     at async a (/app/.next/server/pages/user.js:1:11253)
openlibry  |     at async eZ (/app/.next/server/pages/user.js:1:8232)
openlibry  |     at async e3 (/app/node_modules/next/dist/compiled/next-server/pages.runtime.prod.js:31:594)
openlibry  |     at async doRender (/app/node_modules/next/dist/server/base-server.js:1425:30)
openlibry  |     at async cacheEntry.responseCache.get.routeKind (/app/node_modules/next/dist/server/base-server.js:1599:28)
openlibry  |     at async NextNodeServer.renderToResponseWithComponentsImpl (/app/node_modules/next/dist/server/base-server.js:1507:28) {
openlibry  |   code: 'P2023',
openlibry  |   clientVersion: '5.15.1',
openlibry  |   meta: {
openlibry  |     modelName: 'User',
openlibry  |     message: "Conversion failed: Value 18974892374 does not fit in an INT column, try migrating the 'id' column type to BIGINT"
openlibry  |   }
openlibry  | }
openlibry  |  ⨯ PrismaClientKnownRequestError:
openlibry  | Invalid `prisma.user.findMany()` invocation:
openlibry  |
openlibry  |
openlibry  | Inconsistent column data: Conversion failed: Value 18974892374 does not fit in an INT column, try migrating the 'id' column type to BIGINT
openlibry  |     at In.handleRequestError (/app/node_modules/@prisma/client/runtime/library.js:122:6877)
openlibry  |     at In.handleAndLogRequestError (/app/node_modules/@prisma/client/runtime/library.js:122:6211)
openlibry  |     at In.request (/app/node_modules/@prisma/client/runtime/library.js:122:5919)
openlibry  |     at async l (/app/node_modules/@prisma/client/runtime/library.js:127:11167)
openlibry  |     at async a (/app/.next/server/pages/user.js:1:11253)
openlibry  |     at async eZ (/app/.next/server/pages/user.js:1:8232)
openlibry  |     at async e3 (/app/node_modules/next/dist/compiled/next-server/pages.runtime.prod.js:31:594)
openlibry  |     at async doRender (/app/node_modules/next/dist/server/base-server.js:1425:30)
openlibry  |     at async cacheEntry.responseCache.get.routeKind (/app/node_modules/next/dist/server/base-server.js:1599:28)
openlibry  |     at async NextNodeServer.renderToResponseWithComponentsImpl (/app/node_modules/next/dist/server/base-server.js:1507:28) {
openlibry  |   code: 'P2023',
openlibry  |   clientVersion: '5.15.1',
openlibry  |   meta: {
openlibry  |     modelName: 'User',
openlibry  |     message: "Conversion failed: Value 18974892374 does not fit in an INT column, try migrating the 'id' column type to BIGINT"
openlibry  |   }
openlibry  | }
openlibry  | ERROR in getAllUsers:  PrismaClientKnownRequestError:
openlibry  | Invalid `prisma.user.findMany()` invocation:
openlibry  |
openlibry  |
openlibry  | Inconsistent column data: Conversion failed: Value 18974892374 does not fit in an INT column, try migrating the 'id' column type to BIGINT
openlibry  |     at In.handleRequestError (/app/node_modules/@prisma/client/runtime/library.js:122:6877)
openlibry  |     at In.handleAndLogRequestError (/app/node_modules/@prisma/client/runtime/library.js:122:6211)
openlibry  |     at In.request (/app/node_modules/@prisma/client/runtime/library.js:122:5919)
openlibry  |     at async l (/app/node_modules/@prisma/client/runtime/library.js:127:11167)
openlibry  |     at async a (/app/.next/server/pages/user.js:1:11253)
openlibry  |     at async eZ (/app/.next/server/pages/user.js:1:8232)
openlibry  |     at async e3 (/app/node_modules/next/dist/compiled/next-server/pages.runtime.prod.js:31:594)
openlibry  |     at async doRender (/app/node_modules/next/dist/server/base-server.js:1425:30)
openlibry  |     at async cacheEntry.responseCache.get.routeKind (/app/node_modules/next/dist/server/base-server.js:1599:28)
openlibry  |     at async NextNodeServer.renderToResponseWithComponentsImpl (/app/node_modules/next/dist/server/base-server.js:1507:28) {
openlibry  |   code: 'P2023',
openlibry  |   clientVersion: '5.15.1',
openlibry  |   meta: {
openlibry  |     modelName: 'User',
openlibry  |     message: "Conversion failed: Value 18974892374 does not fit in an INT column, try migrating the 'id' column type to BIGINT"
openlibry  |   }
openlibry  | }
openlibry  |  ⨯ PrismaClientKnownRequestError:
openlibry  | Invalid `prisma.user.findMany()` invocation:
openlibry  |
openlibry  |
openlibry  | Inconsistent column data: Conversion failed: Value 18974892374 does not fit in an INT column, try migrating the 'id' column type to BIGINT
openlibry  |     at In.handleRequestError (/app/node_modules/@prisma/client/runtime/library.js:122:6877)
openlibry  |     at In.handleAndLogRequestError (/app/node_modules/@prisma/client/runtime/library.js:122:6211)
openlibry  |     at In.request (/app/node_modules/@prisma/client/runtime/library.js:122:5919)
openlibry  |     at async l (/app/node_modules/@prisma/client/runtime/library.js:127:11167)
openlibry  |     at async a (/app/.next/server/pages/user.js:1:11253)
openlibry  |     at async eZ (/app/.next/server/pages/user.js:1:8232)
openlibry  |     at async e3 (/app/node_modules/next/dist/compiled/next-server/pages.runtime.prod.js:31:594)
openlibry  |     at async doRender (/app/node_modules/next/dist/server/base-server.js:1425:30)
openlibry  |     at async cacheEntry.responseCache.get.routeKind (/app/node_modules/next/dist/server/base-server.js:1599:28)
openlibry  |     at async NextNodeServer.renderToResponseWithComponentsImpl (/app/node_modules/next/dist/server/base-server.js:1507:28) {
openlibry  |   code: 'P2023',
openlibry  |   clientVersion: '5.15.1',
openlibry  |   meta: {
openlibry  |     modelName: 'User',
openlibry  |     message: "Conversion failed: Value 18974892374 does not fit in an INT column, try migrating the 'id' column type to BIGINT"
openlibry  |   }
openlibry  | }
![Bildschirmfoto_2024-06-21_21-08-09](https://github.com/jzakotnik/openlibry/assets/71968880/887b596c-651c-4ab3-9912-d4f52d15c4ad)

`

**Reply by jzakotnik** [reply]
Ok der Fehler ist eindeutig und nachstellbar, danke, Issue #102 . Das ist wirklich tricky, weil BigInt von `next.js` nicht serialisiert werden kann, da muss man sich was überlegen 👍  
Übrigens, falls es eine Liste der User gibt, das wäre via API viel einfacher das initial automatisiert anzulegen. Ich denke ChatGPT könnte die Liste in eine entsprechende Sequenz von `curl` Kommandos umwandeln.

**Reply by fb4thxUall** [reply]
Hi,

zumindest für uns ist kein BigInt nötig, konnte das jetzt final abklären. Normales Int genügt. Dann würde es zumindest für mich genügen, es bei der aktuellen Version zu belassen. Würde nur den potentiellen Fehler im Code im Sinne der Robustheit abfangen, sodass bei zu große Zahlen nicht gleich alles komplett abschmiert.

Weitere Detailtests mache ich in Kürze.

Danke und LG

**Reply by jzakotnik** [reply]
Danke, ja ich war schon ein wenig ratlos. BigInt ist in node schon recht störrisch, z.B. kann man es nicht einfach so serialisieren und vom Server in den Browser übergeben. Mein Vorschlag wäre eher in der Eingabe ein Limit einzubauen.

**Comment by fb4thxUall** [comment]
Alles erledigt
